torch/_inductor/codecache.py

```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+
+
+@dataclasses.dataclass
+class TypedStorageMetadataHolder:
+    storage_metadata: str
+
+
+class FxGraphCachePickler(pickle.Pickler):
+    """
+    Custom pickler to customize the pickling of some objects, only for the purpose
+    of computing a hash for keying into the FxGraphCache. Some objects don't pickle
+    and/or contain attributes that vary between runs, and we want to capture the
+    data that allow us to compute a safe and stable hash.
+    """
+
+    dispatch_table = copyreg.dispatch_table.copy()
+
+    # Without custom pickling, we get:
+    # AttributeError: Can't pickle local object 'WeakValueDictionary.__init__.<locals>.remove'
+    dispatch_table[torch._subclasses.fake_tensor.FakeTensor] = _reduce_fake_tensor
+
+    # The _cdata attribute is not stable across runs.
+    dispatch_table[torch.storage.TypedStorage] = _reduce_typed_storage
+
+    @staticmethod
+    def dumps(obj) -> bytes:
+        """
+        Pickle an object using the FxGraphCachePickler.
+        """
+        with io.BytesIO() as stream:
+            pickler = FxGraphCachePickler(stream)
+            pickler.dump(obj)
+            return stream.getvalue()
+
+
+@dataclasses.dataclass
+class OrderedSetHolder:
+    items: List[Any]
+
+
+class FxGraphHashDetails:
+    """
+    Object to capture all the details for a compiled FX graph relevant to computing
+    a safe and stable cache key.
+    """
+
+    # Excluded kwargs param that are not stable between runs
+    EXCLUDED_KWARGS = ["graph_id"]
+
+    def __init__(self, fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> None:
+        self.fx_args = fx_args
+
+        # Order kwargs so hashing is stable to changes in kwarg order.
+        self.fx_kwargs = {}
+        for k in sorted(fx_kwargs):
+            if k not in self.EXCLUDED_KWARGS:
+                if type(fx_kwargs[k]) is set:
+                    # Special case to handle set params. Python sets can't be
+                    # ordered, so sort the elements and store them in a proxy.
+                    self.fx_kwargs[k] = OrderedSetHolder(sorted(fx_kwargs[k]))
+                else:
+                    self.fx_kwargs[k] = fx_kwargs[k]
+
+        self.torch_version = torch.__version__
+        self.triton_version = triton.runtime.version_key()
+        self.inductor_config = config.save_config()  # type: ignore[attr-defined]
+
+
+def compiled_fx_graph_hash(fx_args: List[Any], fx_kwargs: Dict[str, Any]) -> str:
+    """
+    Generate a unique hash of the FX graph for caching.
+    """
+    details = FxGraphHashDetails(fx_args, fx_kwargs)
+    serialized_data = FxGraphCachePickler.dumps(details)
+    return (
+        "f"
+        + base64.b32encode(hashlib.sha256(serialized_data).digest())[:51]
+        .decode("utf-8")
+        .lower()
+    )
+
+
+class FxGraphCache:
+    """
+    Supports caching and reusing compiled Fx graphs.
+    """
+
+    # TODO(slarsen): Investigate whether it's beneficial to store compiled graphs
```

Comment:
The first impl from @ani300 was also keeping the data in an in-memory cache, but I can't convince myself that it leads to any noticeable perf improvement. Since @eellison notes a problem when keeping the reference alive anyway, I propose we keep the impl simpler and I'll later look into whether having an in-memory component makes any real difference.